File: LangouAssistant.Core/Privacy/SensitiveContextPolicy.cs
namespace LangouAssistant.Core.Privacy;

public sealed record ContextDescriptor(
    string ProcessName,
    string WindowTitle,
    bool IsPasswordField,
    bool IsSecureDesktop,
    bool IsProtectedWindow,
    string ScreenLabels = "");

public sealed record PrivacyDecision(bool AllowCapture, bool AllowTextUpload, string Reason)
{
    public static PrivacyDecision Allow() => new(true, true, "普通聊天窗口");
    public static PrivacyDecision Block(string reason) => new(false, false, reason);
}

public static class SensitiveContextPolicy
{
    private static readonly string[] BlockedProcessFragments =
    [
        "1password",
        "bitwarden",
        "keepass",
        "lastpass",
        "dashlane",
        "authenticator",
        "credentialui",
        "consent",
        "secur32",
    ];

    private static readonly string[] BlockedTitleFragments =
    [
        "密码",
        "口令",
        "验证码",
        "动态码",
        "支付",
        "付款",
        "确认付款",
        "收银台",
        "银行卡",
        "信用卡",
        "借记卡",
        "银行",
        "金融",
        "安全中心",
        "安全验证",
        "身份验证",
        "转账",
        "汇款",
        "windows security",
        "password",
        "passcode",
        "payment",
        "confirm payment",
        "checkout",
        "wallet",
        "security verification",
        "identity verification",
        "verification code",
        "one-time password",
        "enter pin",
        "bank card",
        "credit card",
        "debit card",
        "money transfer",
    ];

    public static PrivacyDecision Evaluate(ContextDescriptor context)
    {
        ArgumentNullException.ThrowIfNull(context);

        if (context.IsPasswordField)
        {
            return PrivacyDecision.Block("密码输入框禁止采集");
        }

        if (context.IsSecureDesktop)
        {
            return PrivacyDecision.Block("系统安全桌面禁止采集");
        }

        if (context.IsProtectedWindow)
        {
            return PrivacyDecision.Block("受保护窗口禁止采集");
        }

        var process = context.ProcessName.Trim().ToLowerInvariant();
        if (BlockedProcessFragments.Any(process.Contains))
        {
            return PrivacyDecision.Block("密码管理或系统安全程序禁止采集");
        }

        var visibleSecurityContext =
            $"{context.WindowTitle}\n{context.ScreenLabels}".Trim().ToLowerInvariant();
        if (BlockedTitleFragments.Any(visibleSecurityContext.Contains))
        {
            return PrivacyDecision.Block("支付、银行或安全页面禁止采集");
        }

        return PrivacyDecision.Allow();
    }
}

File: production/android/app/src/main/java/com/osfans/trime/langou/privacy/SensitiveContextPolicy.kt
/*
 * SPDX-FileCopyrightText: 2026 Langou Input Method contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package com.osfans.trime.langou.privacy

import android.text.InputType
import java.util.Locale

data class ContextSignals(
    val inputType: Int,
    val packageName: String,
    val screenLabels: String = "",
    val secureWindow: Boolean = false,
)

object SensitiveContextPolicy {
    private val blockedPackageFragments =
        setOf(
            "alipay",
            "unionpay",
            "tenpay",
            ".bank",
            ".icbc",
            ".cmb",
            ".ccb",
            ".boc",
            "wallet",
            "keepass",
            "bitwarden",
            "onepassword",
            "1password",
            "passwordmanager",
        )

    private val sensitiveScreenPattern =
        Regex(
            "支付密码|交易密码|付款码|收银台|银行卡|信用卡|借记卡|安全验证|身份验证|" +
                "验证码|动态码|转账|汇款|确认付款|password|passcode|payment|" +
                "confirm payment|checkout|wallet|security verification|" +
                "identity verification|verification code|one-time password|" +
                "enter pin|bank card|credit card|debit card|money transfer",
            RegexOption.IGNORE_CASE,
        )

    fun canCollect(signals: ContextSignals): Boolean {
        if (signals.secureWindow || isPassword(signals.inputType)) return false
        val normalizedPackage = signals.packageName.lowercase(Locale.ROOT)
        if (blockedPackageFragments.any(normalizedPackage::contains)) return false
        if (sensitiveScreenPattern.containsMatchIn(signals.screenLabels)) return false
        return true
    }

    private fun isPassword(inputType: Int): Boolean {
        val inputClass = inputType and InputType.TYPE_MASK_CLASS
        val variation = inputType and InputType.TYPE_MASK_VARIATION
        return when (inputClass) {
            InputType.TYPE_CLASS_TEXT ->
                variation in
                    setOf(
                        InputType.TYPE_TEXT_VARIATION_PASSWORD,
                        InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD,
                        InputType.TYPE_TEXT_VARIATION_WEB_PASSWORD,
                    )
            InputType.TYPE_CLASS_NUMBER ->
                variation == InputType.TYPE_NUMBER_VARIATION_PASSWORD
            else -> false
        }
    }
}
